### PR TITLE
Make Layer.toImage() accept bounds rather than size

### PR DESCRIPTION
**Pull request by tvolkert**
It allows callers to account for things like shadows that paint
outside the bounds of the widget.

https://github.com/flutter/flutter/issues/16859

### CODE REVIEW

**Comment by tvolkert** [comment]
This includes https://github.com/flutter/flutter/pull/17126

File: packages/flutter/lib/src/rendering/layer.dart [hunk 1]
```diff
@@ -519,8 +519,9 @@ class OffsetLayer extends ContainerLayer {
 
   /// Capture an image of the current state of this layer and its children.
   ///
-  /// The returned [ui.Image] has uncompressed raw RGBA bytes, in the
-  /// dimensions [logicalSize] multiplied by the [pixelRatio].
+  /// The returned [ui.Image] has uncompressed raw RGBA bytes, will be offset
+  /// by [bounds.topLeft], and have dimensions of [bounds.size] multiplied by
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/layer.dart]: unfortunately dartdoc won't like this

File: packages/flutter/lib/src/rendering/layer.dart [hunk 2]
```diff
@@ -532,20 +533,23 @@ class OffsetLayer extends ContainerLayer {
...
-  Future<ui.Image> toImage(Size logicalSize, {double pixelRatio: 1.0}) async {
+  Future<ui.Image> toImage(Rect bounds, {double pixelRatio: 1.0}) async {
+    assert(bounds != null);
     assert(pixelRatio != null);
     final ui.SceneBuilder builder = new ui.SceneBuilder();
-    final Matrix4 transform = new Matrix4.diagonal3Values(pixelRatio, pixelRatio, 1.0);
-    transform.translate(-offset.dx, -offset.dy, 0.0);
+    final Matrix4 transform = new Matrix4.identity();
+    transform.translate(bounds.left, bounds.top);
+    transform.translate(-offset.dx, -offset.dy);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/layer.dart]: in theory you could save a couple of cycles by merging these calls

File: packages/flutter/lib/src/rendering/view.dart [hunk 3]
```diff
@@ -185,7 +185,7 @@ class RenderView extends RenderObject with RenderObjectWithChildMixin<RenderBox>
   }
 
   @override
-  Rect get paintBounds => Offset.zero & size;
+  Rect get paintBounds => Offset.zero & (size * configuration.devicePixelRatio);
```
[the remark this answers is not in the document]
Comment (tvolkert) [reply packages/flutter/lib/src/rendering/view.dart]: Done

**Comment by tvolkert** [comment]
All other comments addressed - merging on green

**Comment by tvolkert** [comment]
Merging on red due to flakes (pub flake and https://github.com/flutter/flutter/issues/17137)

**Comment by andreidiaconu** [comment]
I think this issue might be related https://github.com/flutter/flutter/issues/17686

**Pull request merged**